### ISSUE TITLE
Feature request: Custom layers

### ISSUE THREAD

**Issue by noobco** [UltimateHackingKeyboard/agent#222]
It seems to me that the layers are limited to the mod, fn and mouse layer.

It also seems like scancodes are grouped. e.g. it's being assumed that if you press `1` then it's being assumed that `Shift + 1` will definitely be a `!`.
However this is not the case for a lot of keyboard layouts.

I don't know a whole lot about keyboards so I do naively assume that the keyboard will send scancodes to my computer and my computer will decide based on the scancode and the keyboard layout set on the computer what keycode was meant.

If that's the case I'd like to be able to use a shift layer which will not actually emit shift to the host system but instead make keys from another layer available to me so that I can configure `Shift + 1` to actually be `%` on a system that's actually using the qwerty keyboard layout.

In other words I want to be able to add custom layers.

**Use case example**
For example I use the programmer Dvorak so the key that would be a `1` on the Qwerty layout should be a `&` and it should be a '%' when pressed with the shift modifier.
Furthermore I even have custom `Alt-Gr +` maps on the software layer.
Poorly I can't use a computer of anyone without adding the programmer Dvorak layout first and even then all my custom maps don't work and the programmer Dvorak layout is not even available on Windows by default so it's even worse over there.

**Comment by mondalaci** [comment]
Scancodes are grouped given the nature of the USB protocol. Let's say Shift + 1 is pressed. What really happens behind the scenes, the keyboard sends the Shift modifier, and the Scancode of the 1 key. It doesn't send a specific character, merely a scancode and a modifier. Then the Operating system translates the scancode and the modifier to a character according to the keyboard settings of the OS.

It would be possible to make the UHK check the pressed modifiers and emit a different scancode accordingly, but it would vastly complicate the application, the configuration process, and it'd be confusing for the user for a number of reasons:

1. Agent doesn't have knowledge of the current OS keymap and assumes US English keymap. So it displays `!` for Shift+1, but it's actually `'` for the Hungarian keymap for example.
2. Let's say we have per modifier layers. Then what happens when two modifers get pressed at the same time? It's not clear which would compose with the others or change the scancode.
3. We already have a quite complicated configuration scheme of keymaps, layers, and keys. Complicating it further would be quite overwhelming from a usability standpoint. According to the feedback we received a very small number of users would actually use this feature.

According to my knowledge, there is no keyboard that natively supports the above scenario because of these issues. I'm not sure whether we will ever support it, but at least the opportunity is there because anyone can hack the firmware.
